src/target/riscv/riscv.c

```diff
@@ -462,11 +462,76 @@ static void trigger_from_breakpoint(struct trigger *trigger,
 	trigger->unique_id = breakpoint->unique_id;
 }
 
-static int maybe_add_trigger_t1(struct target *target,
-		struct trigger *trigger, uint64_t tdata1)
+static bool match_napot_checker(struct trigger *trigger, riscv_reg_t *tdata2)
+{
+	riscv_reg_t addr = trigger->address;
+	riscv_reg_t size = trigger->length;
+	if (size > 1 &&
+		(size & (size - 1)) == 0 &&
+		(addr & (size - 1)) == 0) {
+		if (tdata2)
+			*tdata2 = addr | ((size - 1) >> 1);
+		return true;
+	}
+	return false;
+}
+
+static int find_trigger(struct target *target, int type, bool chained, int *idx)
 {
 	RISCV_INFO(r);
 
+	int found;
+	int cnt = 0;
+	int num = chained ? 2 : 1;
+
+	for (unsigned i = 0; i < r->trigger_count; i++) {
+		if (r->trigger_unique_id[i] == -1) {
+			int t = r->trigger_type[i];
+			if (type == t) {
+				found = (++cnt == num);
+				if (found) {
+					if (idx)
+						*idx = i - (num - 1);
+					return ERROR_OK;
+				}
+				continue;
```

Comment:
Please define the `found` variable close to the first place of use. Also use `bool` and a more descriptive name.

Avoid the use of pre-increment and post-increment in expressions. Please put it to a separate statement, where possible.

```suggestion
				num_found++;
				bool done = (num_required == num_found);
				if (done) {
					/* Found num_required consecutive free triggers - success, done. */
					if (idx)
						*idx = i - (num - 1);
					return ERROR_OK;
				}
				/* Found a trigger but need more consecutive ones */
				continue;
```